Cookbook/Proposals/Danger-Integration.md

```diff
@@ -0,0 +1,195 @@
+* Warn if the PR doesn't have the author assigned to it
+* Report SwiftLint warnings in the PR (though it would require to move the execution of Danger after the `test_pr` job, and besides, as of today those SwiftLint violations are reported as errors anyway)
+* Warn when the code introduce a new `// swiftlint:disable` usage in the codebase (especially outside of tests), to invite reviewers to ensure that this SwiftLint exception is justified
+* ...
+
+Those rules could also have been picked as initial candidate for the first rule we implement in Danger, but the rule about the PR size seems more straightforward to implement as a good way to test the initial usage of Danger on our process.
+
+All the rules mentionned in the list above will get their own separate proposal, so that we can discuss their benefits, the exact conditions to trigger each rule, and alternatives considered for each of them separately.
+
+_Note: I am hoping to write proposals for some of those additional rules soon, so that we could discuss them altogether in the upcoming proposal meeting alongside this very proposal for Danger, if time allows. I'll update the list above with links as I publish those other proposals_
+
+### Different conditions for PR size Rule
+
+The rule propose to make the condition based on the "overall lines of code added/removed in the diff" (that's the value returned by `git.lines_of_code` as documented in Danger docs). Which means it checks the value of `insertions` + `deletions`. That means that a PR which adds 20 lines and remove 30 lines will be seen as a size of 50.
+
+We could instead pick one of these other alternatives:
+
+* Define the size of a PR by `abs(insertions - deletions)`, so that a PR adding 30 lines and removing 20 will be seen as of size 10. But since there's no guarantee that those 30 lines added are _replacing_ the 20 lines removed, as they could also be completely unrelated, we don't think it is a reliable metric
+
+* Use a ponderation of `A * insertions + B * deletions`, typically with `A > B > 0`, arguing that it's faster to review deleted code than added code. I'm not convinced that it's really the case – or worth going to such detail – but we might consider trying it in a future PR if we see that Danger warns too often about false positives
+
+* Ignore some files when computing the PR size, considering they don't play a significant role in estimating the PR _review time and complexity_. Here are some file types that we might consider ignoring when computing `insertions` and `deletions`: Snapshot files, Localization files, Assets (like images), ... (Note that this would complexify the implementation of the rule though, as we'd need to compute the number of additions and deletions ourselves by analyzing the diff file by file in the `Dangerfile` instead of direclty using the `insertions` and `deletions` metrics provided by the metadata)
```

Comment:
> Here are some file types that we might consider ignoring when computing `insertions` and `deletions`: **Snapshot files**, Localization files, Assets (like images), ... 

Definite yes for snapshots! I've seen PRs where most of the changes were related to those files, e.g.

```
# just snapshot changes
git log -1 --shortstat 03cf31d839466c0817a4d6640545aac5b85a4663 -- '*.png'
 102 files changed, 204 insertions(+), 204 deletions(-)

# all changes
git log -1 --shortstat 03cf31d839466c0817a4d6640545aac5b85a4663 -- '*.png'
 106 files changed, 326 insertions(+), 234 deletions(-)
```

It would even help with assessing real size/potential complexity of the PR if this calculated number was always posted, not just on warnings.

> (Note that this would complexify the implementation of the rule though, as we'd need to compute the number of additions and deletions ourselves by analyzing the diff file by file in the `Dangerfile` instead of direclty using the `insertions` and `deletions` metrics provided by the metadata)

Agree that we may want to introduce tweaks to the original rule later. btw I checked and it looks like there's already a nice method for this - [`info_for_file`](https://github.com/danger/danger/blob/master/lib/danger/danger_core/plugins/dangerfile_git_plugin.rb#L138-L148).